### PR TITLE
Add NGSI-LD multi-measures support

### PR DESCRIPTION
**Pull request by jason-fox**
- split NGSI-v1, NGSI-v2 and NGSI-LD into separate files
- Move common REST functions into restUtils.js
- ensure that ids are URNS (adding urn:ngsi-ld: where necessary)
- Use Batch Upsert endpoint throughout ( `/ngsi-ld/v1/entityOperations/upsert/`) instead of PATCH update
- Update test expectations
- Split NGSI v2 and NGSI-LD registrations

### CODE REVIEW

**Comment by fgalan** [comment]
PR #843 has been merged and base branch changed to feature/842_ngsi_ld in this PR.

So I understand you should upgrade jason-fox:feature/ngsi-ld-multimeasure with telefonicaid:feature/842_ngsi_ld

File: lib/services/devices/registrationUtils.js [hunk 1]
```diff
@@ -317,15 +314,18 @@ function sendUnregistrationsNgsiLD(deviceData, callback) {
...
-        return deviceService.executeWithSecurity(
+        return utils.executeWithSecurity(
             options,
             deviceData,
-            createRegistrationHandlerNgsi2(true, deviceData, callback));
+            createRegistrationHandlerNgsi2(true, deviceData, callback)
+        );
     }
 
+    //console.error(JSON.stringify(options.json, null, 4));
```
Comment (fgalan) [inline lib/services/devices/registrationUtils.js]: Leftover comment?

File: lib/services/common/genericMiddleware.js [hunk 2]
```diff
@@ -97,7 +97,7 @@ function getLogLevel(req, res, next) {
 /**
  * Ensures the request type is one of the supported ones.
  */
-function ensureType(req, res, next) {
+function ensureType(req, res, next) {
```
Comment (fgalan) [inline lib/services/common/genericMiddleware.js]:
```suggestion
function ensureType(req, res, next) {
```

**Comment by fgalan** [comment]
I think the PR is ok (except for a couple of very minor comments easy to fix)

However, regarding:

> split NGSI-v1, NGSI-v2 and NGSI-LD into separate files

I like the refactor to move separate parts into -NGSI-LD.js -NGSI-v1.js and- NGSI-v2.js files in different parts. However, I wonder about the usual question: which impact could have this change into other existing PRs? Could you analyze it a little bit and provide feedback, pls?

File: lib/services/devices/registrationUtils.js (hunk 1 above)
Comment (jason-fox) [reply lib/services/devices/registrationUtils.js]: Yes, but - the multiple instances of `console.error`  (commented out) are being used to help generate the responses for the tests.

When the final PR is processed I'll remove them all.

File: lib/services/common/genericMiddleware.js (hunk 2 above)
Comment (jason-fox) [reply lib/services/common/genericMiddleware.js]: Not sure how that whitespace occured.

**Comment by jason-fox** [comment]
> split NGSI-v1, NGSI-v2 and NGSI-LD into separate files

This is very simple - All of the _split_ `/devices`, `/ngsi` and `/northBound` files are to do with **NGSI Interactions only** - these are all defined by the relevant specifications, so there is zero likelihood that another PR will be amending them.  It will have **NO* impact on the other PRs. The split has not changed the functionality of the underlying code (since none of the v1 or v2 tests has been modified) - its merely cut-paste rearrangement.

Going forward,  reducing the file size means that it is even less likely that a clash will occur - besides which  it is practically impossible to read a single 2000+ line file with interleaving implementations of three  near duplicates of each function with only fractional differences between them. [sigh]

**Comment by jason-fox** [comment]
Note that things like https://github.com/telefonicaid/iotagent-node-lib/pull/835  can touch files like `deviceService.js`, but this is **common** functionality which applies to  v1 v2 and LD and therefore the relevant functions remain unchanged in the root file.

File: lib/services/devices/registrationUtils.js (hunk 1 above)
Comment (fgalan) [reply lib/services/devices/registrationUtils.js]: Noted in the accumulative PR - https://github.com/telefonicaid/iotagent-node-lib/pull/849#issuecomment-594666173

NTC with regards this PR

File: lib/services/common/genericMiddleware.js (hunk 2 above)
Comment (fgalan) [reply lib/services/common/genericMiddleware.js]: Fixed in cbf6cef

**Pull request merged**